### PR TITLE
Connect users in tracked mail tasks

### PR DESCRIPTION
**Pull request by fross123**
I tweaked one of the logs for the tracker because it was failing a test. Hopefully it still logs what you need, I apologize if it doesn't.

Also, the feature discussed in #112. It works well in my pretty limited situation. I wound up adding a setting and only matching if settings.TODO_MATCH_USERS is true to make sure that current users would not be affected unnecessarily.

Thanks!

### CODE REVIEW

**Review by multun — CHANGES_REQUESTED** [review]
Thanks for your contribution :+1:

It looks good to me overall, I've requested a few changes.
Out of curiosity, are you using the mail tracker in production somewhere? If so, are you happy with the experience?

File: todo/mail/consumers/tracker.py [hunk 1]
```diff
@@ -65,7 +68,7 @@ def parse_references(task_list, references):
     if answer_thread is None:
         logger.info("no answer thread found in references")
     else:
-        logger.info("found an answer thread: %d", answer_thread)
+        logger.info(f"found an answer thread: {answer_thread}")
```
[the remark this answers is not in the document]
Comment (bernd-wechner) [reply todo/mail/consumers/tracker.py:71]: That link is broken. I'd be interested to know the reasoning behind a preference against f-strings.

[the remark this answers is not in the document]
Comment (shacker) [reply todo/mail/consumers/tracker.py:71]: @bernd-wechner The link works for me. The article makes a good point, though I have to see we use f-strings in logging at work and have never been bitten by that issue, which I consider an edge case.

**Comment by shacker** [comment]
@fross123 - @multun Did a code review on this last April, and I was just wondering if you were going to be able to follow up on his comments?

File: todo/mail/consumers/tracker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bernd-wechner) [reply todo/mail/consumers/tracker.py:71]: Cool. that site doesn't use DNSEC and blocked here, so I used a proxy. The point the it makes is valid for high volume logging. The rate at which todo is used here (or likely ever to be on a given site) makes the noted issues irrelevant beyond simply being better practice in general for logging (as some of it is high volume and frequency or logging things expensive to calculate). But all good, I now understand the concern.

**Comment by shacker** [comment]
@fross123 Definitely still interested in having the PR, if you feel like it's close and you're willing!

File: todo/mail/consumers/tracker.py [hunk 2]
```diff
@@ -124,10 +127,18 @@ def insert_message(task_list, message, priority, task_title_format):
 
     with transaction.atomic():
         if best_task is None:
+            user = None
```
[the remark this answers is not in the document]
Comment (fross123) [reply todo/mail/consumers/tracker.py:130]: Absolutely!

File: todo/mail/consumers/tracker.py [hunk 3]
```diff
@@ -124,10 +127,18 @@ def insert_message(task_list, message, priority, task_title_format):
...
+                try:
+                    user = get_user_model().objects.get(email=parseaddr(message_from)[1])
+                except get_user_model().DoesNotExist:
+                    user = None
+
             best_task = Task.objects.create(
                 priority=priority,
                 title=format_task_title(task_title_format, message),
                 task_list=task_list,
+                created_by=user,
```
[the remark this answers is not in the document]
Comment (fross123) [reply todo/mail/consumers/tracker.py:141]: Ooh yeah! Will do this.

File: README.md [hunk 4]
```diff
@@ -261,6 +261,15 @@ TODO_MAIL_TRACKERS = {
 }
 ```
 
+Optionally, you can match incoming users with their accounts.
+
+Ex: If a user emails the test_tracker, and also is a registered User in your application, the user will be added to "created_by" for the task.
+
+Settings:
+```python
+TODO_MATCH_USERS = True # Set to True if you would like to match users. If you do not have authentication setup, do not set this to True.
```
[the remark this answers is not in the document]
Comment (fross123) [reply README.md:270]: Totally hear you on the external authentication. However, I am a little unclear on what the setting should be when we want to use this function, could you expand on that? Where would `get_django_users_by_email` need to be and what should it do?

Also, I will change the name of the option!

File: README.md [hunk 5]
```diff
@@ -261,6 +261,15 @@ TODO_MAIL_TRACKERS = {
 }
 ```
 
+Optionally, you can match incoming users with their accounts.
```
[the remark this answers is not in the document]
Comment (fross123) [reply README.md:264]: Love it, thank you!

**Comment by fross123** [comment]
Sorry about the delay in this folks. I've pushed some updates, let me know what you think. Wasn't too sure about the change requested to the setting, so I will update when we discuss that further.

Thanks!

File: todo/mail/consumers/tracker.py [hunk 6]
```diff
@@ -149,3 +154,18 @@ def tracker_consumer(
         except Exception:
             # ignore exceptions during insertion, in order to avoid
             logger.exception("got exception while inserting message")
+
+
+def match_user(email):
+    """ This function takes an email and checks for a registered user."""
+
+    if not settings.TODO_MAIL_USER_MAPPER:
```
Comment (shacker) [inline todo/mail/consumers/tracker.py:162]: Nice.

File: todo/tests/test_tracker.py [hunk 7]
```diff
@@ -58,3 +58,59 @@ def test_tracker_task_creation(todo_setup, django_user_model):
     Comment.objects.get(
         task=task, body__contains="test3 content", email_message_id="<c@example.com>"
     )
+
+def test_tracker_email_match(todo_setup, django_user_model, settings):
+    """
+    Ensure that a user is added to new lists when sent from registered email
```
Comment (shacker) [inline todo/tests/test_tracker.py:64]: Thanks for adding a test!

**Review by shacker — APPROVED** [review]
LGTM, and apologies for the lengthy response delay. Approved and merging. Cheers!

**Pull request merged**